Core/Systems/CutsceneSystem/CutscenePlayer.cs

```diff
@@ -0,0 +1,201 @@
+				Player.controlUp = false;
+				Player.controlDown = false;
+				Player.controlJump = false;
+				Player.controlHook = false;
+				Player.controlInv = false;
+				Player.controlUseItem = false;
+				Player.controlUseTile = false;
+			}
+		}
+	}
+
+	internal class CutsceneSystem : ModSystem
+	{
+		/// <summary>
+		/// This handles the UI 'fadeout' (in reality, the game draws again over the UI with opacity)
+		/// </summary>
+		/// <param name="spriteBatch"></param>
+		public override void PostDrawInterface(SpriteBatch spriteBatch)
+		{
+			CutscenePlayer mp = Main.LocalPlayer.GetModPlayer<CutscenePlayer>();
+
+			if (mp.InCutscene || mp.fadeTimer > 0)
+			{
+				spriteBatch.Draw(Main.screenTarget, Vector2.Zero, Color.White * (mp.fadeTimer / 60f));
+
+				// Hack to redraw the rich text box, this will likely be the only UI we want here, right?
+				if (UILoader.GetUIState<RichTextBox>().Visible)
+					UILoader.GetUIState<RichTextBox>().Draw(spriteBatch);
+			}
+		}
+	}
+
+	internal static class CutsceneExtensions
+	{
+		/// <summary>
+		/// Sets the player's active cutscene to a new instance of the provided type
+		/// </summary>
+		/// <typeparam name="T">The type of cutscene to begin</typeparam>
+		/// <param name="player">The player to start the cutscene for</param>
+		public static void ActivateCutscene<T>(this Player player) where T : Cutscene
+		{
+			player.GetModPlayer<CutscenePlayer>().SetActiveCutscene<T>();
+		}
+
+		/// <summary>
+		/// Sets the player's active cutscene to a new instance of the provided type
+		/// 
+		/// Typically only used by the net sync
+		/// </summary>
+		/// <param name="player">The player to start the cutscene for</param>
+		/// <param name="fullName">The fully qualified name of the cutscene type</param>
+		public static void ActivateCutscene(this Player player, string fullName)
+		{
+			player.GetModPlayer<CutscenePlayer>().SetActiveCutscene(fullName);
+		}
+
+		/// <summary>
+		/// Deactivates this player's current active cutscene if there is one
+		/// </summary>
+		/// <param name="player">The player to end the cutscene for</param>
+		public static void DeactivateCutscene(this Player player)
+		{
+			player.GetModPlayer<CutscenePlayer>().DeactivateCutscene();
+		}
+
+		/// <summary>
+		/// Checks if the player is in a cutscene of the given type. Check against
+		/// the abstract Cutscene type to check for any cutscene.
+		/// </summary>
+		/// <typeparam name="T">The type of cutscene to check for</typeparam>
+		/// <param name="player">The player to check the cutscene status of</param>
+		/// <returns>If the player is in that cutscene or not</returns>
+		public static bool InCutscene<T>(this Player player) where T : Cutscene
+		{
+			CutscenePlayer mp = player.GetModPlayer<CutscenePlayer>();
+
+			if (mp?.activeCutscene is T)
+				return true;
+
+			return false;
+		}
+	}
+
+	[Serializable]
+	public class CutscenePacket : Module
+	{
+		readonly byte whoAmI;
+		readonly string fullType;
```

Comment:
careful with strings, if a packet goes above 1508 bytes(including header) it gets split on most networks which has a big hit to latency when using TCP like terraria does. though anything between 0 and 1508 bytes has identical performance -- which is why we should bulkify some packets though I think we can do that at the very end once multiplayer is fully supported

Reply:
With how sparsely these are sent I dont think this will become an issue, though I will keep this in mind in the future. The alternative here would be to preload cutscenes into a dictionary like other similar types, though this would also lead to some annoyance with resetting those instances on the client